### ISSUE TITLE
[XLA] SPMD partitioner clamps a partitioned `dynamic-slice` per shard instead of globally

### ISSUE THREAD

**Issue by snarang181** [openxla/xla#49122]
## Summary

For a `dynamic-slice` whose result is partitioned along the sliced dimension and whose start index is out of range, the SPMD partitioner emits `partition_id * shard_size + clamp(raw_start, 0, dim − slice_size)`, i.e. it clamps the raw start against the whole operand and then adds the shard offset. `DynamicSlice` semantics clamp the *global* start first (here 7 → 2) and each result shard then reads its own piece of that window (starts 2 and 5). The emitted starts are 5 and 5, so both shards return the same last columns. Unsharded compilation and NumPy are correct. Similar to #49118 (dynamic-update-slice). 

## Reproducer (CPU only, 2 host devices)

```python
import os
os.environ["XLA_FLAGS"] = "--xla_force_host_platform_device_count=2"
os.environ["JAX_PLATFORMS"] = "cpu"
import numpy as np, jax, jax.numpy as jnp
from jax import lax
from jax.sharding import NamedSharding, PartitionSpec as P

mesh = jax.make_mesh((2,), ("x",), devices=jax.devices("cpu")[:2])
cols = NamedSharding(mesh, P(None, "x"))   # shard the sliced (column) axis
rep  = NamedSharding(mesh, P())

def f(x, start):
    return lax.dynamic_slice(x, [jnp.int32(0), start], (2, 6))

x = jnp.arange(16, dtype=jnp.int32).reshape(2, 8)
start = jnp.int32(7)                        # out of range: clamps to 8 - 6 = 2 -> columns 2..7
expected = np.asarray(x)[:, 2:8]

uns = np.asarray(jax.jit(f)(x, start))
sh  = np.asarray(jax.jit(f, in_shardings=(cols, rep), out_shardings=cols)(x, start))
print("unsharded:", uns[0].tolist(), "ok" if np.array_equal(uns, expected) else "WRONG")
print("sharded:  ", sh[0].tolist(),  "ok" if np.array_equal(sh, expected) else "WRONG")
```

```
unsharded: [2, 3, 4, 5, 6, 7] ok
sharded:   [5, 6, 7, 5, 6, 7] WRONG
```
## Agent Analysis

Post-partitioner HLO for the result-shard slice:

```
%multiply = %partition-id * 3
%clamp    = clamp(0, %raw_start, 5)        ; clamps against the 8-wide operand (8 - 3)
%start    = %multiply + %clamp
ROOT dynamic-slice(%global_value, 0, %start), dynamic_slice_sizes={2,3}
```

For raw start 7 the clamp gives 5 on both partitions, and the surviving dynamic-slice's own clamp into `[0, 8 − 3]` keeps 5 for both. The correct order is `clamp(raw_start, 0, 8 − 6) = 2`, then `2 + partition_id * 3` → `[2, 5]`. Code: the partitioned-dimension dynamic-slice path in `SpmdPartitioningVisitor::HandleDynamicSlice`, `xla/service/spmd/spmd_partitioner.cc`.

**Comment by penumallas1261** [comment]
Internal reproducer submitted in cl/985244532.
